audio/drivers/sdl_audio.c

```diff
@@ -30,25 +30,76 @@
 
 #include "../audio_driver.h"
 #include "../../verbosity.h"
+#include "retro_assert.h"
+
+/**
+ * We need the SDL_{operation}AudioDevice functions for microphone support,
+ * but those were introduced in SDL 2.0.0
+ * (according to their docs).
+ * Some legacy build platforms are stuck on 1.x.x,
+ * so we have to accommodate them.
+ * That comes in the form of stub implementations of the missing functions
+ * that delegate to the non-Device versions.
+ *
+ * Only three platforms (as of this writing) are stuck on SDL 1.x.x,
+ * so it's not a big deal to exclude mic support from them.
+ **/
+#if HAVE_SDL2
+#define SDL_DRIVER_MIC_SUPPORT 1
+#define SDL_DRIVER_DEVICE_FUNCTIONS 1
+#else
+typedef Uint32 SDL_AudioDeviceID;
+
+/** Compatibility stub that defers to SDL_PauseAudio. */
+#define SDL_PauseAudioDevice(dev, pause_on) SDL_PauseAudio(pause_on)
+
+/** Compatibility stub that defers to SDL_LockAudio. */
+#define SDL_LockAudioDevice(dev) SDL_LockAudio()
+
+/** Compatibility stub that defers to SDL_UnlockAudio. */
+#define SDL_UnlockAudioDevice(dev) SDL_UnlockAudio()
+
+/** Compatibility stub that defers to SDL_CloseAudio. */
+#define SDL_CloseAudioDevice(dev) SDL_CloseAudio()
+
+/* Macros for checking audio format bits that were introduced in SDL 2 */
+#define SDL_AUDIO_MASK_BITSIZE       (0xFF)
+#define SDL_AUDIO_MASK_DATATYPE      (1<<8)
+#define SDL_AUDIO_MASK_ENDIAN        (1<<12)
+#define SDL_AUDIO_MASK_SIGNED        (1<<15)
+#define SDL_AUDIO_BITSIZE(x)         (x & SDL_AUDIO_MASK_BITSIZE)
+#define SDL_AUDIO_ISFLOAT(x)         (x & SDL_AUDIO_MASK_DATATYPE)
+#define SDL_AUDIO_ISBIGENDIAN(x)     (x & SDL_AUDIO_MASK_ENDIAN)
+#define SDL_AUDIO_ISSIGNED(x)        (x & SDL_AUDIO_MASK_SIGNED)
+#define SDL_AUDIO_ISINT(x)           (!SDL_AUDIO_ISFLOAT(x))
+#define SDL_AUDIO_ISLITTLEENDIAN(x)  (!SDL_AUDIO_ISBIGENDIAN(x))
+#define SDL_AUDIO_ISUNSIGNED(x)      (!SDL_AUDIO_ISSIGNED(x))
+#endif
 
 typedef struct sdl_audio
 {
 #ifdef HAVE_THREADS
    slock_t *lock;
    scond_t *cond;
 #endif
-   fifo_buffer_t *buffer;
```

Comment:
Some of these fields were renamed because when I was implementing mic support as part of the audio driver I needed to differentiate input and output.